smart-connector-rest-server/src/main/java/eu/knowledge/engine/rest/api/impl/ProactiveApiServiceImpl.java

```diff
@@ -127,6 +129,141 @@ public void scAskPost(
 			return;
 		}
 
+		try {
+			var askFuture = kb.ask(knowledgeInteractionId, recipientAndBindingSetObject.recipient,
+					recipientAndBindingSetObject.bindingSet);
+
+			askFuture.thenAccept(askResult -> {
+
+				LOG.debug("AskResult received, resuming async response: {}", askResult);
+				List<AskExchangeInfo> infos = askResult.getExchangeInfoPerKnowledgeBase().stream()
+						.map(aei -> new AskExchangeInfo().bindingSet(this.bindingSetToList(aei.getBindings()))
+								.knowledgeBaseId(aei.getKnowledgeBaseId().toString())
+								.knowledgeInteractionId(aei.getKnowledgeInteractionId().toString())
+								.exchangeStart(Date.from(aei.getExchangeStart()))
+								.exchangeStart(Date.from(aei.getExchangeStart()))
+								.initiator(toInitiatorEnumAsk(aei.getInitiator()))
+								.exchangeEnd(Date.from(aei.getExchangeEnd())).status(aei.getStatus().toString())
+								.failedMessage(aei.getFailedMessage()))
+						.collect(Collectors.toList());
+
+				LOG.info("Bindings in result is {}", askResult.getBindings());
+				LOG.info("Knowledge gaps in result is {}", askResult.getKnowledgeGaps());
+				AskResult ar = new AskResult().bindingSet(this.bindingSetToList(askResult.getBindings()))
+						.knowledgeGaps(this.knowledgeGapsToList(askResult.getKnowledgeGaps()))
+						.exchangeInfo(infos);
+
+				asyncResponse.resume(Response.status(Status.OK).entity(ar).build());
+			});
+
+		} catch (URISyntaxException | InterruptedException | ExecutionException e) {
+			LOG.trace("", e);
+			var response = new ResponseMessage();
+			response.setMessageType("error");
+			response.setMessage("Something went wrong while sending a POST or while waiting on the REACT.");
+			asyncResponse.resume(Response.status(Status.INTERNAL_SERVER_ERROR).entity(response).build());
+		} catch (IllegalArgumentException e) {
+			LOG.trace("", e);
+			var response = new ResponseMessage();
+			response.setMessageType("error");
+			response.setMessage(e.getMessage());
+			asyncResponse.resume(Response.status(Status.BAD_REQUEST).entity(response).build());
+		}
+	}
+
+	private List<List<String>> knowledgeGapsToList(Set<KnowledgeGap> knowledgeGaps) {
+		List<List<String>> listKnowledgeGaps = new ArrayList<List<String>>(knowledgeGaps.size());
+		knowledgeGaps.forEach((kg) -> {
+			List<String> listKnowledgeGap = new ArrayList<String>();
+			kg.forEach((tp) -> {
+				listKnowledgeGap.add(tp.toString());
```

Comment:
We need to make sure the `tp.toString()` method actually returns valid SPARQL 1.1 triple patterns (that can be processed further if necessary), because currently [it doesn't](https://github.com/TNO/knowledge-engine/blob/rest_api_with_knowledge_gaps/reasoner/src/main/java/eu/knowledge/engine/reasoner/api/TriplePattern.java#L94-L96). Typically, `toString()` is only meant for development purposes, so maybe we could do something like this here:

```
String tpString = FmtUtils.stringForTriple(tp.asTriple(), new PrefixMappingZero());
listKnowledgeGap.add(tpString);
```

This uses the Apache Jena internal `FmtUtils.stringForTriple(...)` method for creating a representation of a triple that follows the official SPARQL 1.1 specification.

This does, however, require an additional method in the [`TriplePattern`](https://github.com/TNO/knowledge-engine/blob/rest_api_with_knowledge_gaps/reasoner/src/main/java/eu/knowledge/engine/reasoner/api/TriplePattern.java) class:

```
public Triple asTriple() {
   return Triple.create(this.getSubject(), this.getPredicate(), this.getObject());
}
```

Reply:
Agree...actually, this was a "bug" in the sense that the returned knowledge gaps were not correct! I made the proposed changes